lang/cs/Org.Apache.REEF.IMRU.Tests/TestActiveContextManager.cs

```diff
@@ -0,0 +1,304 @@
+            activeContextManager.Add(new MyActiveContext(ContextIdPrefix + "3"));
+
+            add = () => activeContextManager.Add(new MyActiveContext(ContextIdPrefix + "4"));
+            Assert.Throws<IMRUSystemException>(add);
+        }
+
+        /// <summary>
+        /// Test remove a failed evaluator which has two contexts associated.
+        /// In current IMRU driver, assume there is only one context associated to the IFailedEvalutor
+        /// </summary>
+        [Fact]
+        public void TestRemoveFailedEvaluatorWithTwoContexts()
+        {
+            var activeContextManager = new ActiveContextManager(5);
+            for (int i = 0; i < 5; i++)
+            {
+                var c = new MyActiveContext(ContextIdPrefix + i);
+                activeContextManager.Add(c);
+            }
+            IList<string> contextIds = new List<string>();
+            contextIds.Add(ContextIdPrefix + "3");
+            contextIds.Add(ContextIdPrefix + "4");
+            var e = new MyFailedEvaluator(contextIds);
+            Action remove = () => activeContextManager.RemoveFailedContextInFailedEvaluator(e);
+            Assert.Throws<IMRUSystemException>(remove);
+        }
+
+        /// <summary>
+        /// Test remove a failed evaluator which has a context but it doesn't exist.
+        /// </summary>
+        [Fact]
+        public void TestRemoveFailedEvaluatorWithNoExistsContexts()
+        {
+            var activeContextManager = new ActiveContextManager(5);
+            for (int i = 0; i < 5; i++)
+            {
+                var c = new MyActiveContext(ContextIdPrefix + i);
+                activeContextManager.Add(c);
+            }
+
+            IList<string> contextIds = new List<string>();
+            contextIds.Add(ContextIdPrefix + "5");
+            var e = new MyFailedEvaluator(contextIds);
+            Action remove = () => activeContextManager.RemoveFailedContextInFailedEvaluator(e);
+            Assert.Throws<IMRUSystemException>(remove);
+        }
+
+        /// <summary>
+        /// Test remove a failed evaluator which has no context associated.
+        /// The scenario may happen when an evaluator failed but context has not created yet. 
+        /// </summary>
+        [Fact]
+        public void TestRemoveFailedEvaluatorWithNoContext()
+        {
+            var activeContextManager = new ActiveContextManager(5);
+            for (int i = 0; i < 5; i++)
+            {
+                var c = new MyActiveContext(ContextIdPrefix + i);
+                activeContextManager.Add(c);
+            }
+
+            var e = new MyFailedEvaluator(null);
+            activeContextManager.RemoveFailedContextInFailedEvaluator(e);
+            Assert.Equal(5, activeContextManager.NumberOfActiveContexts);
+        }
+
+        /// <summary>
+        /// An implementation of IFailedEvaluator for testing
+        /// </summary>
+        private class MyFailedEvaluator : IFailedEvaluator
```

Comment:
Please use `NSubstitute` instead of creating private classes, for all private classes in this file.